trickler/scales.py

```diff
@@ -344,11 +344,91 @@ def _stable_unstable(self, line):
         self.resolution = self.resolution_map[self.unit]
         # Update memcache values.
         self._update_memcache()
+        
+        
+class USSolidScale(SerialScale):
+    """Class for controlling a U.S. Solid USS-DBS81-110G scale."""
+
+    def __init__(self, config, port='/dev/ttyUSB0', baudrate=9600, timeout=0.1, **kwargs):
+        """Only overriding this to provide scale specific constructor arguments."""
+        super().__init__(config=config, port=port, baudrate=baudrate, timeout=timeout, **kwargs)
+
+    @classmethod
+    @property
+    def unit_map(cls):
+        """Mapping of self.unit keys to string units of weight as used by the scale."""
+        return {
+            'gn': cls.Units.GRAINS,
+            'g': cls.Units.GRAMS,
+         }
+
+    @classmethod
+    @property
+    def resolution_map(cls):
+        """Map self.units to matching resolutions with decimal.Decimal values."""
+        return {
+            cls.Units.GRAINS: decimal.Decimal('0.01'),
+            cls.Units.GRAMS: decimal.Decimal('0.001'),
+        }
+
+    # Note(eric): There is no documentation on how to do this for this scale.
+    def change_unit(self):
+        """Changes the unit of weight on the scale."""
+        logging.debug('changing weight unit on scale from: %r', self.unit)
+        # Send Mode button command.
+        self._serial.write(b'U\r\n')
+        # Sleep 1s and wait for change to take effect.
+        time.sleep(1)
+        # Run update fn to set latest values.
+        self.update()
+
+    def update(self):
+        """Read from the serial port and update an instance of this class with the most recent values."""
+        handlers = {
+            '+': self._stable_unstable,
+            '-': self._stable_unstable,
+            None: noop,
+        }
+
+        # Note: The input buffer can fill up, causing latency. Clear it before reading.
+        self._serial.reset_input_buffer()
+        raw = self._serial.readline()
+        logging.debug(raw)
+        try:
+            # Remove trailing newline characters, then decode from bytestring into unicode.
+            line = raw.rstrip(b'\r\n').decode('utf-8')
+        except UnicodeDecodeError:
+            logging.debug('Could not decode bytes to unicode.')
+        else:
+            # Pull the first character (positive or negative, on this scale)
+            prefix = line[0:1]
+            # Get a handler function based on the prefix
+            handler = handlers.get(prefix, noop)
+            # Run the function to handle the input.
+            handler(line)
+
+    def _stable_unstable(self, line):
+        """Update the scale when status is stable or unstable."""
+        # Push the latest reading into the internal list.
+        self._readings.append(line)
+        # Update internal stability bit.
+        self._check_stability()
+        # Store the numeric weight from the scale reading.
+        weight = line[0:8]
+        self.weight = decimal.Decimal(weight)
```

Comment:
Because your scale has a space character (`" "`) between the positive/negative prefix and the numeric value, it's causing an issue with `decimal.Decimal`.

Here's an example, using some of the output values you provided from your scale:

```
>>> decimal.Decimal('+ 145.203')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
decimal.InvalidOperation: [<class 'decimal.ConversionSyntax'>]


>>> decimal.Decimal('- 145.203')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
decimal.InvalidOperation: [<class 'decimal.ConversionSyntax'>]
```

When you remove that space character however, it works fine:

```
>>> decimal.Decimal('+145.203')
Decimal('145.203')
```

One way to handle this is to just remove the space characters from this string.

```
>>> decimal.Decimal('- 145.203'.replace(' ', ''))
Decimal('-145.203')

>>> decimal.Decimal('+ 145.203'.replace(' ', ''))
Decimal('145.203')
```

To do that, in your code, I'd suggest this change:

```
weight = weight.replace(' ', '')
self.weight = decimal.Decimal(weight)
```